### PR TITLE
[FIVE-6] Crear Modulo de Login

### PR DESCRIPTION
**Pull request by gonzalorom**
**Descripcion:**
[FIVE-6] Modulo Login y Regitrarse en basado en Cognito.

**Realizado:**
- Instalados paquetes Amazon.AspNetCore.Identity.Cognito.
- Agregados los servicios SignIn, SignUp, User y Configuration.
- Agregadas las clases SignInDTO y SignUpDTO con sus respectivos models.
- Creado servicio para extraccion de credenciales (AWS) desde DB.
- Se agrega filtro de Autentificacion.
- Se agrega frontend: Pagina Sign in y Sign up.
- Se agrega React Router
- Se crea Private Route
- Se migra a Hooks.
- Se agrega validacion de usuario registrado.
- Se crea logout.

[FIVE-6]: https://makingsense.atlassian.net/browse/FIVE-6

### CODE REVIEW

**Comment by gonzalorom** [comment]
-Se agrega frontend: Pagina Sign in y Sign up.
-Se agrega React Router
-Se migra a Hooks.
-Se agrega validacion de usuario registrado.
-Se crea logout.

File: FiveRockingFingers/FRF.Web/ClientApp/src/commons/NavMenu.tsx [hunk 1]
```diff
@@ -1,5 +1,5 @@
 import * as React from 'react';
-import { ProSidebar, Menu, MenuItem, SubMenu, SidebarHeader, SidebarFooter, SidebarContent } from 'react-pro-sidebar';
+import { ProSidebar, Menu, SidebarHeader, SidebarContent } from 'react-pro-sidebar';
 import 'react-pro-sidebar/dist/css/styles.css';
```
Comment (egutbrod) [inline FiveRockingFingers/FRF.Web/ClientApp/src/commons/NavMenu.tsx:3]: Keep imports sorted

File: FiveRockingFingers/FRF.Web/ClientApp/src/commons/NavMenuItem.tsx [hunk 2]
```diff
@@ -1,18 +1,25 @@
...
 const NavMenuItem = (props: { project: Project }) => {
     return (
         <div>
             <SubMenu title={`Proyecto: ${props.project.name}`}>
-                <MenuItem>Artefactos</MenuItem>
-                <MenuItem>Equipo</MenuItem>
-                <MenuItem>Presupuesto</MenuItem>
+                <Box><MenuItem active={false}>
+                    <p>{`Creador: ${props.project.owner}`}</p>
+                    <p>{`Cliente: ${props.project.client}`}</p>
```
Comment (egutbrod) [inline FiveRockingFingers/FRF.Web/ClientApp/src/commons/NavMenuItem.tsx:12]: This code is part of [FIVE-4]

File: FiveRockingFingers/FRF.Web/ClientApp/src/commons/SnackbarError.tsx [hunk 3]
```diff
@@ -0,0 +1,24 @@
+import React from "react";
+import Snackbar from '@material-ui/core/Snackbar';
```
Comment (egutbrod) [inline FiveRockingFingers/FRF.Web/ClientApp/src/commons/SnackbarError.tsx:2]: Keep imports sorted

File: FiveRockingFingers/FRF.Web/ClientApp/src/components/Home.tsx [hunk 4]
```diff
@@ -1,4 +1,4 @@
-import * as React from 'react';
+import React from 'react';
```
Comment (egutbrod) [inline FiveRockingFingers/FRF.Web/ClientApp/src/components/Home.tsx:1]: Keep imports sorted

File: FiveRockingFingers/FRF.Web/ClientApp/src/components/auth/Login.tsx [hunk 5]
```diff
@@ -0,0 +1,117 @@
+import React from "react";
+import { Form, FormGroup, Label, Col, Row } from "reactstrap";
+import { Paper, Button, FormControlLabel, Checkbox, TextField } from "@material-ui/core";
+import { useHistory } from "react-router-dom";
+import { useUserContext } from "./contextLib";
+import axios from "axios"
```
Comment (egutbrod) [inline FiveRockingFingers/FRF.Web/ClientApp/src/components/auth/Login.tsx:6]: Keep imports sorted

File: FiveRockingFingers/FRF.Web/ClientApp/src/components/auth/Signup.tsx [hunk 6]
```diff
@@ -0,0 +1,174 @@
+import React from 'react';
+import { Form, FormGroup, Label, Row } from 'reactstrap';
+import { Paper, Button, TextField, Box } from '@material-ui/core';
+import { useHistory } from 'react-router-dom';
+import { useUserContext } from "./contextLib";
+import axios from 'axios'
```
Comment (egutbrod) [inline FiveRockingFingers/FRF.Web/ClientApp/src/components/auth/Signup.tsx:6]: Keep imports sorted

File: FiveRockingFingers/FRF.Web/ClientApp/src/index.tsx [hunk 7]
```diff
@@ -1,17 +1,16 @@
 import 'bootstrap/dist/css/bootstrap.css';
-
 import * as React from 'react';
 import * as ReactDOM from 'react-dom';
-import { createBrowserHistory } from 'history';
```
Comment (egutbrod) [inline FiveRockingFingers/FRF.Web/ClientApp/src/index.tsx:5]: Keep imports sorted

File: FiveRockingFingers/FRF.Web/ClientApp/src/router/Routers.tsx [hunk 8]
```diff
@@ -0,0 +1,18 @@
+import React from 'react';
+import { Route, Switch, Redirect } from "react-router-dom";
+import Home from '../components/Home';
+import Login from '../components/auth/Login';
```
Comment (egutbrod) [inline FiveRockingFingers/FRF.Web/ClientApp/src/router/Routers.tsx:4]: Keep imports sorted

File: FiveRockingFingers/FRF.Web/ClientApp/src/router/Routes.tsx [hunk 9]
```diff
@@ -1,13 +1,18 @@
-﻿import React from 'react';
-import { Switch, Route } from 'react-router';
+import React from 'react';
+import { Route, Switch, Redirect } from "react-router-dom";
 import Home from '../components/Home';
```
Comment (egutbrod) [inline FiveRockingFingers/FRF.Web/ClientApp/src/router/Routes.tsx:3]: Keep imports sorted

File: FiveRockingFingers/FRF.Web/Controllers/BaseApiController.cs [hunk 10]
```diff
@@ -3,11 +3,13 @@
 using System.Collections.Generic;
 using System.Linq;
```
Comment (egutbrod) [inline FiveRockingFingers/FRF.Web/Controllers/BaseApiController.cs:4]: Sort usings

File: FiveRockingFingers/FRF.Web/Controllers/SignInController.cs [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+
+namespace FRF.Web.Controllers
+{
+    [AllowAnonymous]
+    [Route("api/[controller]")]
+    [ApiController]
+    public class SignInController : ControllerBase
+    {
+        private readonly ISignInService SignInService;
+        private readonly IMapper Mapper;
```
Comment (egutbrod) [inline FiveRockingFingers/FRF.Web/Controllers/SignInController.cs:18]: Change these private attributes names to _name

File: FiveRockingFingers/FRF.Web/Controllers/SignUpController.cs [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+
+namespace FRF.Web.Controllers
+{
+    [Route("api/[controller]")]
+    [AllowAnonymous]
+    [ApiController]
+    public class SignUpController : ControllerBase
+    {
+        private readonly ISignUpService SignUpService;
+        private readonly IMapper Mapper;
```
Comment (egutbrod) [inline FiveRockingFingers/FRF.Web/Controllers/SignUpController.cs:18]: Change these private attributes names to _name

File: FiveRockingFingers/FRF.Web/Controllers/UserController.cs [hunk 13]
```diff
@@ -0,0 +1,58 @@
...
+using System.Security.Claims;
+using System.Threading.Tasks;
+
+namespace FRF.Web.Controllers
+{
+    [Route("api/[controller]")]
+    [ApiController]
+    public class UserController : ControllerBase
+    {
+        public IUserService UserService { get; set; }
```
Comment (egutbrod) [inline FiveRockingFingers/FRF.Web/Controllers/UserController.cs:15]: Change to:

`private readonly IUserService _userService;`

**Review by egutbrod — CHANGES_REQUESTED** [review]

File: FiveRockingFingers/FRF.Core/Services/ConfigurationService.cs [hunk 14]
```diff
@@ -0,0 +1,50 @@
...
+using Microsoft.Extensions.Configuration;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace FRF.Core.Services
+{
+    public class ConfigurationService : IConfigurationService
+    {
+        public DataAccessContext DataContext { get; set; }
```
[the remark this answers is not in the document]
Comment (gonzalorom) [reply FiveRockingFingers/FRF.Core/Services/ConfigurationService.cs]: Done!

File: FiveRockingFingers/FRF.Core/Services/ConfigurationService.cs [hunk 15]
```diff
@@ -0,0 +1,50 @@
...
+            DataContext = dataContext;
+        }
+
+        
+        public CognitoConfigurationBase GetConfigurationSettings()
+        {
+            try
+            {
+                var resultFromDb = DataContext.ConfigurationSettings.ToDictionary(k => k.Name, v => v.Value);
+                var result = ToObject<CognitoConfigurationBase>(resultFromDb);
```
[the remark this answers is not in the document]
Comment (gonzalorom) [reply FiveRockingFingers/FRF.Core/Services/ConfigurationService.cs:20]: Done!

File: FiveRockingFingers/FRF.Core/Services/ISignInService.cs [hunk 16]
```diff
@@ -0,0 +1,11 @@
+﻿using System;
+using FRF.Core.Models;
```
[the remark this answers is not in the document]
Comment (gonzalorom) [reply FiveRockingFingers/FRF.Core/Services/ISignInService.cs]: Done!

File: FiveRockingFingers/FRF.Core/Services/ISignUpService.cs [hunk 17]
```diff
@@ -0,0 +1,11 @@
+﻿using System;
+using FRF.Core.Models;
```
[the remark this answers is not in the document]
Comment (gonzalorom) [reply FiveRockingFingers/FRF.Core/Services/ISignUpService.cs]: Done!

File: FiveRockingFingers/FRF.Core/Services/SignInService.cs [hunk 18]
```diff
@@ -0,0 +1,41 @@
...
+using FRF.Core.Models;
+using Microsoft.AspNetCore.Identity;
+using System;
+using System.Threading.Tasks;
+
+namespace FRF.Core.Services
+{
+    public class SignInService : ISignInService
+    {
+        private readonly SignInManager<CognitoUser> SignInManager;
```
[the remark this answers is not in the document]
Comment (gonzalorom) [reply FiveRockingFingers/FRF.Core/Services/SignInService.cs]: Done!

File: FiveRockingFingers/FRF.Core/Services/SignUpService.cs [hunk 19]
```diff
@@ -0,0 +1,81 @@
...
+using System.Collections.Generic;
+using System.Net;
+using System.Threading.Tasks;
+
+namespace FRF.Core.Services
+{
+    public class SignUpService : ISignUpService
+    {
+        private readonly CognitoConfigurationBase CognitoBase;
+        private readonly SignInManager<CognitoUser> SignInManager;
```
[the remark this answers is not in the document]
Comment (gonzalorom) [reply FiveRockingFingers/FRF.Core/Services/SignUpService.cs]: Done!

File: FiveRockingFingers/FRF.Core/Services/UserService.cs [hunk 20]
```diff
@@ -0,0 +1,57 @@
+﻿using System;
+using Amazon.Extensions.CognitoAuthentication;
+using Microsoft.AspNetCore.Identity;
```
[the remark this answers is not in the document]
Comment (gonzalorom) [reply FiveRockingFingers/FRF.Core/Services/UserService.cs:4]: Done!

File: FiveRockingFingers/FRF.Core/Services/UserService.cs [hunk 21]
```diff
@@ -0,0 +1,57 @@
...
+using System.Threading.Tasks;
+
+namespace FRF.Core.Services
+{
+    public class UserService : IUserService
+    {
+        private readonly UserManager<CognitoUser> UserManager;
+        private readonly SignInManager<CognitoUser> SignInManager;
```
[the remark this answers is not in the document]
Comment (gonzalorom) [reply FiveRockingFingers/FRF.Core/Services/UserService.cs]: Done!

File: FiveRockingFingers/FRF.DataAccess/DataAccessContext.cs [hunk 22]
```diff
@@ -18,5 +19,9 @@ protected override void OnModelCreating(ModelBuilder builder)
         {
             builder.Entity<ProjectCategory>().HasKey(pc => new { pc.ProjectId, pc.CategoryID });
         }
+        protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
+        {
+            optionsBuilder.UseSqlServer("server=DESKTOP-28EG8E0;database=FiveRockingFingers;trusted_connection=true;");
```
[the remark this answers is not in the document]
Comment (gonzalorom) [reply FiveRockingFingers/FRF.DataAccess/DataAccessContext.cs]: Done!

File: FiveRockingFingers/FRF.Web/ClientApp/src/App.tsx [hunk 23]
```diff
@@ -1,14 +1,30 @@
-import * as React from 'react';
-import { Route, Router, Switch } from 'react-router';
-import { createBrowserHistory } from 'history';
-import './custom.css';
+import React, { useState } from 'react';
+import { UserContext } from "./components/auth/contextLib";
+import { Button } from 'reactstrap';
```
[the remark this answers is not in the document]
Comment (gonzalorom) [reply FiveRockingFingers/FRF.Web/ClientApp/src/App.tsx]: Done!

File: FiveRockingFingers/FRF.Web/ClientApp/src/commons/LoadingButton.tsx [hunk 24]
```diff
@@ -0,0 +1,35 @@
+﻿import React from "react";
+import Fade from "@material-ui/core/Fade";
```
[the remark this answers is not in the document]
Comment (gonzalorom) [reply FiveRockingFingers/FRF.Web/ClientApp/src/commons/LoadingButton.tsx]: Done!

File: FiveRockingFingers/FRF.Web/ClientApp/src/commons/NavMenu.tsx (hunk 1 above)
Comment (gonzalorom) [reply FiveRockingFingers/FRF.Web/ClientApp/src/commons/NavMenu.tsx:6]: Done!

File: FiveRockingFingers/FRF.Web/ClientApp/src/commons/NavMenuItem.tsx (hunk 2 above)
Comment (gonzalorom) [reply FiveRockingFingers/FRF.Web/ClientApp/src/commons/NavMenuItem.tsx]: Done!

File: FiveRockingFingers/FRF.Web/ClientApp/src/commons/SnackbarError.tsx (hunk 3 above)
Comment (gonzalorom) [reply FiveRockingFingers/FRF.Web/ClientApp/src/commons/SnackbarError.tsx]: Done!

File: FiveRockingFingers/FRF.Web/ClientApp/src/components/Home.tsx (hunk 4 above)
Comment (gonzalorom) [reply FiveRockingFingers/FRF.Web/ClientApp/src/components/Home.tsx:1]: Done!

File: FiveRockingFingers/FRF.Web/ClientApp/src/components/auth/Login.tsx (hunk 5 above)
Comment (gonzalorom) [reply FiveRockingFingers/FRF.Web/ClientApp/src/components/auth/Login.tsx]: Done!

File: FiveRockingFingers/FRF.Web/ClientApp/src/components/auth/Signup.tsx (hunk 6 above)
Comment (gonzalorom) [reply FiveRockingFingers/FRF.Web/ClientApp/src/components/auth/Signup.tsx]: Done!

File: FiveRockingFingers/FRF.Web/ClientApp/src/index.tsx (hunk 7 above)
Comment (gonzalorom) [reply FiveRockingFingers/FRF.Web/ClientApp/src/index.tsx:5]: Done!

File: FiveRockingFingers/FRF.Web/ClientApp/src/router/Routers.tsx (hunk 8 above)
Comment (gonzalorom) [reply FiveRockingFingers/FRF.Web/ClientApp/src/router/Routers.tsx]: Done!

File: FiveRockingFingers/FRF.Web/ClientApp/src/router/Routes.tsx (hunk 9 above)
Comment (gonzalorom) [reply FiveRockingFingers/FRF.Web/ClientApp/src/router/Routes.tsx]: Done!

File: FiveRockingFingers/FRF.Web/Controllers/BaseApiController.cs (hunk 10 above)
Comment (gonzalorom) [reply FiveRockingFingers/FRF.Web/Controllers/BaseApiController.cs]: Done!

File: FiveRockingFingers/FRF.Web/Controllers/SignInController.cs (hunk 11 above)
Comment (gonzalorom) [reply FiveRockingFingers/FRF.Web/Controllers/SignInController.cs]: Done!

File: FiveRockingFingers/FRF.Web/Controllers/SignUpController.cs (hunk 12 above)
Comment (gonzalorom) [reply FiveRockingFingers/FRF.Web/Controllers/SignUpController.cs]: Done!

File: FiveRockingFingers/FRF.Web/Controllers/UserController.cs (hunk 13 above)
Comment (gonzalorom) [reply FiveRockingFingers/FRF.Web/Controllers/UserController.cs]: Done!

File: FiveRockingFingers/FRF.Web.Dtos/Users/SignUpDTO.cs [hunk 25]
```diff
@@ -0,0 +1,25 @@
...
+{
+    public class SignUpDTO
+    {
+        [Required]
+        public string Name { get; set; }
+        [Required]
+        public string FamilyName { get; set; }
+        [Required]
+        [EmailAddress]
+        [RegularExpression(@"^((([!#$%&'*+\-/=?^_`{|}~\w])|([!#$%&'*+\-/=?^_`{|}~\w][!#$%&'*+\-/=?^_`{|}~\.\w]{0,}[!#$%&'*+\-/=?^_`{|}~\w]))[@]\w+([-.]\w+)*\.\w+([-.]\w+)*)$")]
```
[the remark this answers is not in the document]
Comment (gonzalorom) [reply FiveRockingFingers/FRF.Web.Dtos/Users/SignUpDTO.cs]: Done!

**Comment by matiasbeckerle** [comment]
This PR is massive hehe. Did you addressed all the requested changes from @egutbrod?

**Comment by gonzalorom** [comment]
Yes, all the request for change from @egutbrod were resolved. I know it is a massive PR.  We are just doing one PR once we finished with the ticket, In this case, the ticket was an epic. This is something that we can improve in the next springs.

**Comment by matiasbeckerle** [comment]
Alright. If everything was addressed then let's move forward by merging it. @egutbrod once you're back from your vacations have a look just in case you want anything else to be updated / improved.

**Review by manquius — CHANGES_REQUESTED** [review]

File: FiveRockingFingers/FRF.Web/Controllers/SignUpController.cs [hunk 26]
```diff
@@ -9,28 +9,36 @@
...
         }
 
+       
         [HttpPost]
         public async Task<ActionResult<string>> SignUp(SignUpDTO signUpDto)
         {
+            /*Clear this after do*/
+            return BadRequest();
+            /********************/
+            /*Uncomment this after do.*/
```
[the remark this answers is not in the document]
Comment (gonzalorom) [reply FiveRockingFingers/FRF.Web/Controllers/SignUpController.cs]: Done.

File: FiveRockingFingers/FRF.Web/Controllers/SignUpController.cs [hunk 27]
```diff
@@ -9,28 +9,36 @@
...
-            _signUpService = signUpService;
+           // _signUpService = signUpService;
             _mapper = mapper;
         }
 
+       
         [HttpPost]
         public async Task<ActionResult<string>> SignUp(SignUpDTO signUpDto)
         {
+            /*Clear this after do*/
```
[the remark this answers is not in the document]
Comment (gonzalorom) [reply FiveRockingFingers/FRF.Web/Controllers/SignUpController.cs]: Done.

File: FiveRockingFingers/FRF.Core/Services/UserService.cs [hunk 28]
```diff
@@ -7,14 +7,17 @@ namespace FRF.Core.Services
 {
     public class UserService : IUserService
     {
-        private readonly SignInManager<CognitoUser> _signInManager;
-        private readonly UserManager<CognitoUser> _userManager;
+        /*Uncomment this after do.*/
```
[the remark this answers is not in the document]
Comment (gonzalorom) [reply FiveRockingFingers/FRF.Core/Services/UserService.cs:10]: Done.

**Pull request merged**